### PR TITLE
Info: Support viewing current media requests and hiding media

### PR DESCRIPTION
**Pull request by mia-pi-git**
Requests from various room staff.
- support /hidetexting requested media once shown (done with using /uhtml)
- support viewing pending requests (done in a page)
- modlog who approves / denies a link 
On another note, makes the comment on /requestshow more viewable by putting it in a box.

### CODE REVIEW

**Comment by Zarel** [comment]
Just use regular HTML, like `!code`...

File: server/chat-commands/info.ts [hunk 1]
```diff
@@ -2509,14 +2507,14 @@ export const commands: ChatCommands = {
...
 		if (request.comment) {
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c|${request.name}|/uhtml ${request.name},${buf}`).update();
```
Comment (Zarel) [inline server/chat-commands/info.ts:2516]:
```suggestion
		room.add(`|c| ${request.name}|/raw ${buf}`).update();
```

**Comment by Zarel** [comment]
What's the use-case for viewing pending requests? Should pending requests just be shown to auth when they join the room?

**Comment by mia-pi-git** [comment]
I would say no to showing upon join, since that might get very spammy. 
The main use case that was given to me is that sometimes people will request and staff will go offline (or the like) and they'll get ignored.

**Comment by Zarel** [comment]
If it gets spammy, they should turn off approvals. Rooms should not say they support approvals and then never approve anything.

**Comment by Zarel** [comment]
Approvals should also refresh at the bottom every few minutes, to prevent this sort of thing.

**Comment by mia-pi-git** [comment]
I mean, you can approve them sure, but it feels like it'd be very annoying to join and immediately get sent several large bright boxes.

**Comment by Zarel** [comment]
If it's annoying, they should turn off approvals. It's much more annoying to the users to have their requests get ignored.

File: server/chat-commands/info.ts [hunk 2]
```diff
@@ -2509,14 +2507,14 @@ export const commands: ChatCommands = {
...
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c| ${request.name}|/raw ${buf}`).update();
+		this.modlog('APPROVESHOW', null, `${request.name} (link: ${request.link})`);
```
Comment (Zarel) [inline server/chat-commands/info.ts:2517]: Do we really need to modlog these? They're already in the roomlog.

Comment (mia-pi-git) [reply server/chat-commands/info.ts:2517]: Modlogging was specifically requested, as it's easier to see there.

**Comment by Zarel** [comment]
Approvals should also preview the image. Your links aren't even clickable! Why force mods to copy/paste a link to see what they're approving?

File: server/chat-commands/info.ts (hunk 2 above)
Comment (Zarel) [reply server/chat-commands/info.ts:2517]: It spams the modlog so much, though. Maybe what we really need is a separate approval log.

Comment (Zarel) [reply server/chat-commands/info.ts:2517]: Yeah, I'm gonna veto modlog.

Comment (Zarel) [reply server/chat-commands/info.ts:2517]: What they want is a roomlog search. Useful feature to build into `/chatlog`. But I don't want it in modlogs.

File: server/chat-commands/info.ts [hunk 3]
```diff
@@ -2509,14 +2509,14 @@ export const commands: ChatCommands = {
...
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c|${request.name}|/uhtml ${request.name},${buf}`).update();
+		this.privateModAction(`${user.name} has approved the display of ${request.name}'s media approval.`);
```
Comment (Zarel) [inline server/chat-commands/info.ts:2519]:
```suggestion
		this.privateModAction(`${user.name} approved showing ${request.name}'s media.`);
```

Comment (mia-pi-git) [reply server/chat-commands/info.ts:2519]: I intentionally made this longer so /approvallog would scrape them easier.

File: server/chat-commands/info.ts [hunk 4]
```diff
@@ -2509,14 +2509,14 @@ export const commands: ChatCommands = {
...
 		if (request.comment) {
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c|${request.name}|/uhtml ${request.name},${buf}`);
```
Comment (whalemer) [inline server/chat-commands/info.ts:2518]: Why is this uhtml?

Comment (Zarel) [reply server/chat-commands/info.ts:2518]: Good question, I thought I told Mia to fix that. Or was that a different thing?

File: server/chat-commands/info.ts (hunk 3 above)
Comment (Zarel) [reply server/chat-commands/info.ts:2519]: You shouldn't add bad grammar to give approvallog an easier time (not to mention that shorter search terms should have better perf). Just add an actual token to search for if it's an issue.

Comment (Zarel) [reply server/chat-commands/info.ts:2519]:
```suggestion
		this.privateModAction(`${user.name} approved showing media from ${request.name}.`);
```

File: server/chat-commands/info.ts [hunk 5]
```diff
@@ -2509,14 +2509,14 @@ export const commands: ChatCommands = {
...
 		if (request.comment) {
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c|${request.name}|/html ${request.name},${buf}`);
```
Comment (Zarel) [inline server/chat-commands/info.ts:2518]: Have you tested this? I'm pretty sure this isn't what you wanted. I'm also pretty sure I already told you the correct way to do this:

https://github.com/smogon/pokemon-showdown/pull/6907#discussion_r444684882

And once again, you marked it as resolved without actually fixing it.

File: server/chat-commands/info.ts [hunk 6]
```diff
@@ -2509,14 +2513,14 @@ export const commands: ChatCommands = {
...
 		if (request.comment) {
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c|${request.name}|/uhtml ${request.name},${buf}`).update();
```
Comment (whalemer) [inline server/chat-commands/info.ts:2522]: you marked this as resolved without actually changing it - i suspect it got lost in your forcepush

Comment (Zarel) [reply server/chat-commands/info.ts:2522]: lol, is this the THIRD time this happened?

Comment (mia-pi-git) [reply server/chat-commands/info.ts]: Sigh, yes it is. 
No more committing from github web.

File: server/chat-commands/info.ts (hunk 3 above)
Comment (Zarel) [reply server/chat-commands/info.ts:2523]: .

File: server/chat-commands/info.ts [hunk 7]
```diff
@@ -2509,14 +2513,14 @@ export const commands: ChatCommands = {
...
 		if (request.comment) {
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
 		}
-		this.addBox(buf);
-		room.update();
+		room.add(`|c|${request.name}|/raw ${buf}`).update();
```
Comment (Zarel) [inline server/chat-commands/info.ts:2522]:
```suggestion
		room.add(`|c|${request.name}|/raw ${buf}`);
```

Comment (Zarel) [reply server/chat-commands/info.ts:2522]: The change you made wasn't the correct one. :|

File: server/chat-commands/info.ts [hunk 8]
```diff
@@ -2528,15 +2532,23 @@ export const commands: ChatCommands = {
...
 
 		room.pendingApprovals!.delete(target);
 		room.sendMods(`|uhtmlchange|request-${target}|`);
-		this.privateModAction(`(${user.name} denied ${target}'s request to display ${link}.)`);
+		this.privateModAction(`(${user.name} denied ${target}'s request to display ${entry.link}.)`);
 	},
 	denyshowhelp: [`/denyshow [user] - Denies the media display request of [user]. Requires: % @ # &`],
 
+	approvallog(target, room, user) {
+		return this.parse(`/sl has approved the display of, ${room.roomid}`);
```
Comment (Zarel) [inline server/chat-commands/info.ts:2545]:
```suggestion
		return this.parse(`/sl approved showing media from, ${room.roomid}`);
```

File: server/chat-commands/info.ts [hunk 9]
```diff
@@ -2647,6 +2659,21 @@ export const pages: PageTable = {
...
+		if (!this.can('mute', null, this.room)) return;
+		if (!room.pendingApprovals) room.pendingApprovals = new Map();
+		if (room.pendingApprovals.size < 1) return `<h2>No pending approvals on ${room.title}</h2>`;
+		let buf = `<div class="pad"><strong>Pending media requests on ${room.title}</strong><hr />`;
+		for (const [userid, entry] of room.pendingApprovals) {
+			buf += `<strong>${entry.name}</strong><div class="infobox">`;
+			buf += `<strong>Requester ID:</strong> ${userid}<br>`;
+			buf += `<strong>Link:</strong> <a href="${entry.link}">${entry.link}</a><br>`;
+			buf += `<strong>Comment:</strong> ${entry.comment}`;
+			buf += `</div><hr/ >`;
```
Comment (Zarel) [inline server/chat-commands/info.ts:2673]:
```suggestion
			buf += `</div><hr />`;
```

File: server/rooms.ts [hunk 10]
```diff
@@ -1285,6 +1285,20 @@ export class BasicChatRoom extends BasicRoom {
...
+			message += `\n|raw|<div class="infobox">`;
+			message += `<details><summary>(Pending media requests: ${this.pendingApprovals.size})</summary>`;
+			for (const [userid, entry] of this.pendingApprovals) {
+				message += `<div class="infobox">`;
+				message += `<strong>Requester ID:</strong> ${userid}<br/ >`;
+				message += `<strong>Link:</strong> <a href="${entry.link}">${entry.link}</a><br/ >`;
+				message += `<strong>Comment:</strong> ${entry.comment ? entry.comment : 'None.'}<br/ >`;
+				message += `<button class="button" name="send" value="/approveshow ${userid}">Approve</button>` +
+				`<button class="button" name="send" value="/denyshow ${userid}">Deny</button></div>`;
+				message += `</div><hr/ >`;
```
Comment (Zarel) [inline server/rooms.ts:1298]:
```suggestion
				message += `</div><hr />`;
```

File: server/rooms.ts [hunk 11]
```diff
@@ -1285,6 +1285,20 @@ export class BasicChatRoom extends BasicRoom {
...
 				`</div>`;
 		}
+		if (this.pendingApprovals && this.pendingApprovals.size > 0) {
+			message += `\n|raw|<div class="infobox">`;
+			message += `<details><summary>(Pending media requests: ${this.pendingApprovals.size})</summary>`;
+			for (const [userid, entry] of this.pendingApprovals) {
+				message += `<div class="infobox">`;
+				message += `<strong>Requester ID:</strong> ${userid}<br/ >`;
+				message += `<strong>Link:</strong> <a href="${entry.link}">${entry.link}</a><br/ >`;
+				message += `<strong>Comment:</strong> ${entry.comment ? entry.comment : 'None.'}<br/ >`;
```
Comment (Zarel) [inline server/rooms.ts:1295]:
```suggestion
				message += `<strong>Requester ID:</strong> ${userid}<br />`;
				message += `<strong>Link:</strong> <a href="${entry.link}">${entry.link}</a><br />`;
				message += `<strong>Comment:</strong> ${entry.comment ? entry.comment : 'None.'}<br />`;
```

File: server/chat-commands/info.ts [hunk 12]
```diff
@@ -2647,6 +2659,21 @@ export const pages: PageTable = {
...
+	approvals(args) {
+		const room = Rooms.get(args[0]) as ChatRoom | GameRoom;
+		if (!this.can('mute', null, this.room)) return;
+		if (!room.pendingApprovals) room.pendingApprovals = new Map();
+		if (room.pendingApprovals.size < 1) return `<h2>No pending approvals on ${room.title}</h2>`;
+		let buf = `<div class="pad"><strong>Pending media requests on ${room.title}</strong><hr />`;
+		for (const [userid, entry] of room.pendingApprovals) {
+			buf += `<strong>${entry.name}</strong><div class="infobox">`;
+			buf += `<strong>Requester ID:</strong> ${userid}<br>`;
+			buf += `<strong>Link:</strong> <a href="${entry.link}">${entry.link}</a><br>`;
```
Comment (Zarel) [inline server/chat-commands/info.ts:2671]:
```suggestion
			buf += `<strong>Requester ID:</strong> ${userid}<br />`;
			buf += `<strong>Link:</strong> <a href="${entry.link}">${entry.link}</a><br />`;
```

File: server/rooms.ts [hunk 13]
```diff
@@ -1285,6 +1285,20 @@ export class BasicChatRoom extends BasicRoom {
 				this.settings.staffMessage.replace(/\n/g, '') +
 				`</div>`;
 		}
+		if (this.pendingApprovals && this.pendingApprovals.size > 0) {
```
Comment (Zarel) [inline server/rooms.ts:1288]:
```suggestion
		if (this.pendingApprovals && this.pendingApprovals.size > 0 && user.can('mute', null, this)) {
```

File: server/rooms.ts [hunk 14]
```diff
@@ -1285,6 +1285,20 @@ export class BasicChatRoom extends BasicRoom {
 				this.settings.staffMessage.replace(/\n/g, '') +
 				`</div>`;
 		}
+		if (this.pendingApprovals && this.pendingApprovals.size > 0 && user.can('mute', null, this)) {
```
Comment (Zarel) [inline server/rooms.ts:1288]:
```suggestion
		if (this.pendingApprovals?.size && user.can('mute', null, this)) {
```

File: server/chat-commands/info.ts [hunk 15]
```diff
@@ -2471,6 +2471,10 @@ export const commands: ChatCommands = {
 			comment: comment,
 		});
 		this.sendReply(`You have requested to show the link: ${link}${comment ? ` (with the comment ${comment})` : ''}.`);
+		const notifyId = `${room.roomid}-rank-driver`;
+		let message = `|tempnotify|${notifyId}|Pending media request!`;
+		message += `|${user.name} has requested to show media in ${room.title}.|new media request`;
```
Comment (Zarel) [inline server/chat-commands/info.ts:2476]:
```suggestion
		const message = `|tempnotify|pendingapprovals|Pending media request!` +
			`|${user.name} has requested to show media in ${room.title}.|new media request`;
```

You don't need the room ID in the notification ID; notifications are already scoped per-room.

File: server/chat-commands/info.ts [hunk 16]
```diff
@@ -2528,15 +2532,23 @@ export const commands: ChatCommands = {
 		target = toID(target);
 		if (!target) return this.parse(`/help denyshow`);
 
-		const link = room.pendingApprovals?.get(target);
-		if (!link) return this.errorReply(`${target} has no pending request.`);
+		const entry = room.pendingApprovals?.get(target);
+		if (!entry) return this.errorReply(`${target} has no pending request.`);
 
 		room.pendingApprovals!.delete(target);
 		room.sendMods(`|uhtmlchange|request-${target}|`);
```
Comment (Zarel) [inline server/chat-commands/info.ts:2538]:
```suggestion
		room.sendMods(`|uhtmlchange|request-${target}|`);
		room.sendRankedUsers(`|tempnotifyoff|pendingapprovals`, '%');
```

(`approveshow` also needs this but it's out of the area I can suggest changes for.)

File: server/chat-commands/info.ts [hunk 17]
```diff
@@ -2509,14 +2513,14 @@ export const commands: ChatCommands = {
...
 			}
 		}
-		buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray"><small>(Requested by ${request.name})</small>`;
+		buf += Utils.html`<br /><div class="infobox"><small>(Requested by ${request.name})</small>`;
 		if (request.comment) {
-			buf += Utils.html`<br /><p style="margin-left:5px;font-size:9pt;color:gray">${request.comment}</p>`;
+			buf += Utils.html`<br />${request.comment}</small></div>`;
 		} else {
-			buf += `</small></p>`;
+			buf += `</small></div>>`;
```
Comment (Zarel) [inline server/chat-commands/info.ts:2520]:
```suggestion
			buf += `</small></div>`;
```

**Comment by Zarel** [comment]
Looks good, thanks!

**Pull request merged**